internal/controllers/catalog_test.go

```diff
@@ -0,0 +1,109 @@
+		var ctx context.Context
+		var opNames = []string{"prometheus", "project-quay"}
+		var reqChan = make(chan string, 4*len(opNames))
+		BeforeEach(func() {
+			ctx = context.Background()
+			mgr, err := ctrl.NewManager(cfg, ctrl.Options{Scheme: sch})
+			Expect(err).To(BeNil())
+
+			var fakeReconciler reconcile.Func = func(_ context.Context, request ctrl.Request) (ctrl.Result, error) {
+				// ignore reconcile requests from other tests
+				if !strings.HasPrefix(request.Name, "operator-test") {
+					reqChan <- request.Name
+				}
+				return ctrl.Result{}, nil
+			}
+
+			err = controllers.SetupWithManager(fakeReconciler, mgr)
+			Expect(err).To(BeNil())
+
+			var mgrCtx context.Context
+			mgrCtx, cancel = context.WithCancel(ctx)
+
+			go func() {
+				err := mgr.Start(mgrCtx)
+				Expect(err).To(BeNil())
+			}()
+
+			for _, p := range opNames {
+				op := &operatorsv1alpha1.Operator{ObjectMeta: metav1.ObjectMeta{Name: p}, Spec: operatorsv1alpha1.OperatorSpec{PackageName: p}}
+				err := cl.Create(ctx, op)
+				Expect(err).To(BeNil())
+			}
+			By("verifying initial reconcile logs for operator creation")
+			var reqNames []string
+			for range opNames {
+				var req string
+				Eventually(reqChan).Should(Receive(&req))
+				reqNames = append(reqNames, req)
+			}
+			Expect(reqNames).To(ConsistOf(opNames))
+		})
+		It("reconciles all affected operators on cluster", func() {
+			By("creating a new catalog")
+			catalog := &catalogd.Catalog{ObjectMeta: metav1.ObjectMeta{Name: "t"}, Spec: catalogd.CatalogSpec{Source: catalogd.CatalogSource{Type: catalogd.SourceTypeImage, Image: &catalogd.ImageSource{}}}}
+			err := cl.Create(ctx, catalog)
+			Expect(err).To(BeNil())
+			By("verifying operator reconcile logs on catalog create")
+			var reqNames []string
+			for range opNames {
+				var req string
+				Eventually(reqChan).Should(Receive(&req))
+				reqNames = append(reqNames, req)
+			}
+			Expect(reqNames).To(ConsistOf(opNames))
+
+			By("updating a catalog")
+			catalog.Spec.Source.Image.Ref = "s"
+			err = cl.Update(ctx, catalog)
+			Expect(err).To(BeNil())
+			By("verifying operator reconcile logs on catalog update")
+			reqNames = []string{}
+			for range opNames {
+				var req string
+				Eventually(reqChan).Should(Receive(&req))
+				reqNames = append(reqNames, req)
+			}
+			Expect(reqNames).To(ConsistOf(opNames))
+
+			By("deleting a catalog")
+			err = cl.Delete(ctx, catalog)
+			Expect(err).To(BeNil())
+			By("verifying operator reconcile logs on catalog delete")
+			reqNames = []string{}
+			for range opNames {
+				var req string
+				Eventually(reqChan).Should(Receive(&req))
+				reqNames = append(reqNames, req)
+			}
+			Expect(reqNames).To(ConsistOf(opNames))
+		})
+		AfterEach(func() {
+			for _, p := range opNames {
+				op := &operatorsv1alpha1.Operator{ObjectMeta: metav1.ObjectMeta{Name: p}, Spec: operatorsv1alpha1.OperatorSpec{PackageName: p}}
+				Expect(cl.Delete(ctx, op)).To(Succeed())
+			}
+			cancel() // stop manager
```

Comment:
We should stop the manager before the clean up. That way we don't attempt any further reconciles based on these deletions (and cause undesired channel send attempts).